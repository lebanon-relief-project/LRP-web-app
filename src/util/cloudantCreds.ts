import { CloudantCredentials } from "../types/Cloudant";

export function getCredentials(): CloudantCredentials {
  const VCAP = JSON.parse(process.env.VCAP_SERVICES);

  if (!VCAP || !VCAP.cloudantNoSQLDB) {
    throw new Error("CLOUDANT_CREDENTIALS missing");
  }

  const apiKey = VCAP.cloudantNoSQLDB[0].credentials.apikey;
  const host = VCAP.cloudantNoSQLDB[0].credentials.url;

  return {
    url: `${host}`,
    apiKey: `${apiKey}`,
  };
}
