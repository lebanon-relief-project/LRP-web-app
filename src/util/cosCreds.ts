import { COS_URL } from "../statics";
import { CosCredentials } from "../types/Cos";

export function getCosCredentials(): CosCredentials {

  const VCAP = JSON.parse(process.env.VCAP_SERVICES);

  if (!VCAP || !VCAP["cloud-object-storage"]) {
    throw new Error("COS credentials missing");
  }
  
  const endpoint = COS_URL;
  const apikey = VCAP["cloud-object-storage"][0].credentials.apikey;
  const resourceInstanceId =
    VCAP["cloud-object-storage"][0].credentials["resource_instance_id"];
  const accessKeyId =
    VCAP["cloud-object-storage"][0].credentials["cos_hmac_keys"][
      "access_key_id"
    ];
  const secretAccessKey =
    VCAP["cloud-object-storage"][0].credentials["cos_hmac_keys"][
      "secret_access_key"
    ];

  return {
    endpoint: endpoint,
    apiKeyId: apikey,
    serviceInstanceId: resourceInstanceId,
    signatureVersion: "v4",
    accessKeyId: accessKeyId,
    secretAccessKey: secretAccessKey,
  };
}
