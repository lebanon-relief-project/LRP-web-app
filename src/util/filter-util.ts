import { FilterType } from "src/types/Filter";

export function sanitizeFilter(unSafeFilter: any): FilterType {
  const schema: Record<keyof FilterType, string> = {
    languages: "string",
  };

  let cleanFilter: FilterType = {};

  Object.keys(schema).forEach((key) => {
    if (unSafeFilter[key]) {
      cleanFilter[key] = unSafeFilter[key];

      if (typeof schema[key] === "string") {
        if (typeof unSafeFilter[key] === "string")
          cleanFilter[key] = [unSafeFilter[key]];
        cleanFilter[key] = cleanFilter[key].map((entry) => {
          return entry.toLowerCase();
        });
      }

      if (typeof schema[key] === "object") {
        cleanFilter[key] = JSON.parse(cleanFilter[key]);
      }
    }
  });

  return cleanFilter;
}
