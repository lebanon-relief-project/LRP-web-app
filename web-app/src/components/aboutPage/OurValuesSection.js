import React, { useEffect, useState, useRef } from "react";
import styled from "styled-components";
import colours from "../../styles/Colours";
import FlashCard from "../FlashCard";
import LoadingSpinner from "../LoadingSpinner";
import { getFlashCards } from "../../services/flashCards.service";
import devices from "../../styles/Devices";
import { useHistory } from "react-router-dom";
import NoCardSelectedModal from "../NoCardSelectedModal";
import { getOurValuesCards } from "../../services/ourValuesCards.service";

const OurValuesSection = (props) => {
  const [noCardSelectedModalVisible, setNoCardSelectedModalVisible] =
    useState(false);

  const [loading, isLoading] = useState(true);

  const [cardWrapperHeight, setCardWrapperHeight] = useState("100%");
  const cardWrapperRef = useRef(null);

  let history = useHistory();
  const [flashCards, setFlashCards] = useState([]);

  const [, /* state */ setState] = useState();

  const fetchFlashCards = async () => {
    let response = await getOurValuesCards();

    isLoading(false);
    setFlashCards(response.cards);
  };

  useEffect(() => {
    fetchFlashCards().catch(() => {
      setState(() => {
        throw new Error("failed to fetch flashCards");
      });
    });
  }, []);

  const handleClick = (e) => {
    e.preventDefault();
    if (sessionStorage.length === 0) {
      setNoCardSelectedModalVisible(true);
    } else {
      history.push("/results");
    }
  };

  //Calculate height of the card container div
  useEffect(() => {
    function setHeight() {
      setCardWrapperHeight(undefined);
      let height = cardWrapperRef?.current?.getBoundingClientRect()?.height;

      if (height > 0) {
        if (
          window.innerWidth >= devices.ipad.replace("px", "") &&
          window.innerWidth <= devices.ipadprolandscape.replace("px", "")
        ) {
          setCardWrapperHeight(height * 0.7);
        } else {
          setCardWrapperHeight(height * 0.5);
        }
      }
    }
    setHeight();
    window.addEventListener("resize", setHeight);
  }, [flashCards]);

  return (
    <>
      <StyledSection>
        {loading ? (
          <LoadingSpinner />
        ) : (
          <Wrapper
            style={{ maxHeight: cardWrapperHeight }}
            ref={cardWrapperRef}
          >
            {flashCards.map((flashCard, index) => {
              return (
                <FlashCard
                  id={`${flashCard._id}`}
                  key={`${flashCard._id}_${index}`}
                  card={flashCard}
                />
              );
            })}
          </Wrapper>
        )}
      </StyledSection>
    </>
  );
};

const StyledSection = styled.section`
  background: inherit;
  margin: 0;
  padding: 0;
  width: 100%;
  display: flex;
  justify-content: center;
  align-items: center;
`;

const Wrapper = styled.div`
  width: 100%;

  max-width: 1440px;

  padding: 0 240px;

  background: inherit;
  display: flex;
  flex-direction: row;
  flex-wrap: wrap;
  justify-content: flex-start;
  align-content: space-between;
  flex-direction: column;

  @media (max-width: ${devices.ipadpro}) {
    padding: 0 18px;
    width: 100%;
  }

  @media (max-width: ${devices.ipad}) {
    justify-content: center;
    max-height: unset !important;
  }
`;

export default OurValuesSection;
