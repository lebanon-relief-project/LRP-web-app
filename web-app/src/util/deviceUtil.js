// Util for use in conditional statements within code (not just css)
export const deviceSize = {
  mobile: 576,
  ipad: 768,
  ipadpro: 1025,
};
