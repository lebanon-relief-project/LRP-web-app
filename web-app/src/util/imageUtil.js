// Converts an ArrayBuffer to image data to be used in a <img />
export const getSvgImageFromArrayBuffer = (bufferData) => {
  return URL.createObjectURL(
    new Blob([new Buffer(bufferData)], { type: "image/svg+xml" })
  );
};
