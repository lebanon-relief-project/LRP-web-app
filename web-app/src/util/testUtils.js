import React from "react";
import { MemoryRouter } from "react-router-dom";
import { render } from "@testing-library/react";

export const renderWithRouter = (component, path = "/") => {
  return render(
    <MemoryRouter initialEntries={[path]}>{component}</MemoryRouter>
  );
};
