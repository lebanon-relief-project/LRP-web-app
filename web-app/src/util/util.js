export const getCardIdsFromSessionStorage = () => {
  let cardIds = [];
  let key;
  const numberOfCards = sessionStorage.length;

  for (let i = 0; i < numberOfCards; i++) {
    key = sessionStorage.key(i);

    // Should regex this to match flashcard id patterns
    if (key !== "null" && key !== "undefined") {
      cardIds.push(key);
    }
  }

  return cardIds;
};

export const getCardTitleFromSessionStorage = (id) => {
  let card = "";
  const numberOfCards = sessionStorage.length;

  for (let i = 0; i < numberOfCards; i++) {
    let key = sessionStorage.key(i);

    if (key === id) {
      card = sessionStorage.getItem(key);
    }
  }

  return card;
};

export const checkIfCardIsInSessionStorage = (id) => {
  let response = false;
  const numberOfCards = sessionStorage.length;

  for (let i = 0; i < numberOfCards; i++) {
    let key = sessionStorage.key(i);

    if (key === id) {
      response = true;
    }
  }

  return response;
};