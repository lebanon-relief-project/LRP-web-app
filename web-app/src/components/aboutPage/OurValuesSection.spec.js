import React from "react";
import { wait } from "@testing-library/react";
import "jest-styled-components";
import { renderWithRouter } from "../../util/testUtils";
import "jest-styled-components";
import { fireEvent } from "@testing-library/dom";
import { act } from "react-dom/test-utils";
import { renderHook } from "@testing-library/react-hooks";
import OurValuesSection from "./OurValuesSection";

jest.mock("../FlashCard", (props) => {
  return (props) => <div>flash card {props.card.title}</div>;
});

jest.mock("../LoadingSpinner", () => {
  return () => <div data-testid="spinner">Loading...</div>;
});

const mockGetOurValuesCards = jest.fn();

mockGetOurValuesCards.mockResolvedValue({
  cards: [
    {
      _id: "random _id",
      title: "random title",
      body: "some card body",
    },
  ],
});

jest.mock("../../services/ourValuesCards.service", () => {
  return {
    getOurValuesCards: () => mockGetOurValuesCards(),
  };
});

jest.mock("../NoCardSelectedModal", () => {
  return () => <div data-testid="modal">No Card Selected Modal</div>;
});

beforeEach(() => {
  jest.clearAllMocks();
});

describe("the OurValuesSection component", () => {
  let container;
  it("should match the snapshot", async () => {
    ({ container } = renderWithRouter(<OurValuesSection />));
    await wait();
    expect(container).toMatchSnapshot();
  });

  it("should show Spinner when our values section is loading", async () => {
    let queryByTestId;
    ({ queryByTestId } = renderWithRouter(<OurValuesSection />));
    expect(queryByTestId("spinner")).toBeTruthy();
    await wait();
    expect(queryByTestId("spinner")).toBeFalsy();
  });

  it("should throw error when database call", async () => {
    mockGetOurValuesCards.mockImplementation(() => {
      throw new Error("ugly error");
    });
    let caughtError = null;
    class ErrorBoundary extends React.Component {
      constructor(props) {
        super(props);
        this.state = { hasError: false };
      }

      componentDidCatch(error) {
        this.setState({ hasError: true });

        caughtError = error;
      }

      render() {
        return !this.state.hasError && this.props.children;
      }
    }

    const wrapper = ({ children }) => <ErrorBoundary>{children}</ErrorBoundary>;

    renderHook(() => OurValuesSection(), { wrapper });

    await wait();

    expect(caughtError).toEqual(new Error("failed to fetch flashCards"));
  });
});
