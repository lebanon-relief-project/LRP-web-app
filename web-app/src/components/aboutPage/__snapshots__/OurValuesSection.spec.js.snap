// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the OurValuesSection component should match the snapshot 1`] = `
.c0 {
  background: inherit;
  margin: 0;
  padding: 0;
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1 {
  width: 100%;
  max-width: 1440px;
  padding: 0 240px;
  background: inherit;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  -webkit-align-content: space-between;
  -ms-flex-line-pack: space-between;
  align-content: space-between;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

@media (max-width:1025px) {
  .c1 {
    padding: 0 18px;
    width: 100%;
  }
}

@media (max-width:768px) {
  .c1 {
    -webkit-box-pack: center;
    -webkit-justify-content: center;
    -ms-flex-pack: center;
    justify-content: center;
    max-height: unset !important;
  }
}

<div>
  <section
    class="c0"
  >
    <div
      class="c1"
    >
      <div>
        flash card 
        random title
      </div>
    </div>
  </section>
</div>
`;
